Seed Driven Cluster Bootstrap

dlex opened the pull request.

andrwng committed:
storage: add node_uuid to storage API

This will be used when we first start a node that has no node UUID.
I exposed this as storage because the UUID uniquely identifies the
contents of the kv-store / data directory.

andrwng committed:
cluster: introduce bootstrap service

This adds the skeleton of an RPC service that will be used for cluster
bootstrap. It includes an initial `cluster_bootstrap_info` RPC type that
is currently empty.

Introducing as a standalone commit to parallelize some workstreams.

andrwng committed:
cluster: encapsulate cluster discovery

This encapsulates some pieces of cluster discovery into its own class. I
added some descriptions on how each method is expected to be
implemented, but I added implementations that match the existing
behavior.

andrwng committed:
application: generate node UUID on startup

Upon first starting up the storage layer, Redpanda will now check to see
if a node UUID exists in the controller keyspace of the key-value store.
If not, it will generate one and persist it to disk, and make it
available on all shards.

This will be used as a unique identifier when registering a node
(particularly important in the future when we auto-assign node IDs).

andrwng committed:
cluster: command to register a UUID

This adds a new controller command to register a node UUID.

It is activated by the existing join RPC endpoint and has the following
semantics:
- Register: all new versions of Redpanda will send their UUID along with
  an optional node ID in their initial join request. If the UUID is not
  in the leader controller's `members_table`, the UUID is registered by
  replicating a `register_node_uuid_cmd`.
- Join: all new version of Redpanda will, after registering, send
  another join request that will actually add the node to the Raft
  group. Old versions of Redpanda will send a join request with no UUID.

andrwng committed:
net: add node_id label to client metrics

RPC `client_probe` metrics currently leverage a labeling scheme defined
by Seastar in which, for a given metric, a set of metric labels can only
be registered once per metric. Currently, the label used is solely based
on the server address associated with a given `rpc::transport`. As such,
we currently cannot start multiple `rpc::transport`s pointed at the same
server.

This functionality could be useful though: consider when a node is
restarted empty with a new node ID. Redpanda currently has a check that
nodes being added to the controller Raft group don't overlap with
existing group members' addresses. But if we were to remove this check,
when the new node _is_ added to the Raft group, each node will try to
create a new `rpc::transport` pointing at the new node, and register
metrics with identical labels to those registered by the old node, and
be met with a Seastar `double_registration` exception.

To enable the above scenario, this commit adds the `node_id` as a label
for client metrics, and aggregates them by this label.

andrwng committed:
features: add flag for node_id_assignment

andrwng committed:
cluster: allow node to join if address is used

andrwng committed:
tests: ducktape test for node ID assignment

andrwng committed:
redpanda: plumb cluster discovery into application

This plumbs the cluster discovery module into `redpanda::application` as
one of the first subsystems initialized during startup.

Cluster discovery will have an initial dependency on the storage
subsystem, to get node UUID, cluster UUID, etc., and on the RPC
subsystem, to serve the RPCs required for the initial bootstrapping
sequence.

The rest of the subsystems will have a dependency on cluster discovery,
which, once automatic node ID assignment is supported, will be what
determines the node ID.

To that end, this commit orders the startup of the storage and RPC
subsystems up front, followed by cluster discovery, followed by the rest
of the subsystems.

This commit only plumbs cluster discovery infrastructure. The cluster
discovery sequence as a whole is unchanged.

dlex committed:
config: add empty_seed_starts_cluster to node_config

Also a corresponding parameter is added to redpanda test fixture

dlex committed:
cluster: adjustments to "node ID assignments"

cluster_discovery works with node config directly
assert message in replicate_and_wait adjusted
comments

dlex committed:
cluster: analysis of seed_servers by cluster_discovery

initial_seed_brokers() now support both ESCB and SDCB modes
determine_node_id() now supports index-based node_ids

dlex committed:
net: unresolved_address got a comparison op

plus some factoring

dlex committed:
cluster: bootstrap controller command handling

cluster_uuid type
New bootstrap controller log command carrying cluster_uuid
New backend to handle cluster bootstrap, w/separate batch type
Persistence of cluster_uuid in kvstore

dlex committed:
cluster: cluster_uuid read from kvstore

dlex committed:
cluster: bootstrap with seed_brokers

Supply clusterDiscovery's initial_seed_brokers list to raft0
as the set of nodes to start the raft0 group with

dlex committed:
cluster: bootstrap command issued

If there is no cluster_uuid in SDCB mode, cluster is initialized
by posting the new bootstrap command to the controller log.

dlex committed:
c/tests: verify single node SDCB mode

cluster_test_fixure now supports the empty_seed_starts_cluster parameter

dlex committed:
cluster: bootstrap command creates default user

bootstrap_backend operates on credentail_storage for that.
In security_frontend, maybe_create_bootstrap_user() is split into two:
get_bootstrap_user_creds_from_env(), and the actual user creation.

dlex commented:
* added implementations in `cluster_discovery` to support local config analysis
* cluster creation by generating a cluster UUID and recording it to controller log
* saving cluster UUID to kvstore
* integration with controller
* unit test to test single node cluster creation in Seed Driven Cluster Bootstrap (SDCB) mode
* the existing tests keep verifying Empty Seed Cluster Bootstrap (ESCB) mode successfully
* the bootstrap command now captures the initial user setting from the env and creates a user like 'create_user' command does

andrwng commented on src/v/cluster/controller.cc:
```diff
@@ -99,13 +99,51 @@ ss::future<> controller::wire_up() {
+    if (!seed_brokers.empty()) {
+        initial_raft0_brokers = std::move(seed_brokers);
+        if (!config::node().empty_seed_starts_cluster())
+            verify_seed_brokers_list(initial_raft0_brokers);
```
nit: might be cleaner to move this logic into `cluster_discovery::initial_seed_brokers()`? Especially since we'll need to fork the logic there based on `empty_seed_starts_cluster` anyway.

andrwng committed:
util: add named_type static constructor from args

It can be convenient to construct a named_type using constructors of the
underlying type.

A simple solution today is to construct the underlying type separately
and then construct the named_type with it, but this spills the
underlying type into call sites. Another approach would be to construct
with the named_type::type constructor, but this can be verbose.

This commit adds a templatized static constructor that callers can pass
constructor args into.

andrwng committed:
config: make node ID optional

This commit makes the node_id config property optional. It doesn't make
any checks that the field is set. A follow-up will change the Redpanda's
behavior when seeing an empty node ID.

andrwng committed:
cluster: unit tests for node ID assignment

andrwng committed:
cluster: auto-assign node ID when empty

This commit adds the ability for Redpanda to assign node ID based on the
node UUID when a node ID isn't set.

andrwng committed:
tests: add option to not configure node ID

andrwng committed:
cluster_discovery: read node ID from kvstore

When the controller starts up, we check to ensure the configured node ID
is identical to that in the kvstore if one exists, persisting one if
not. We should use this node ID at startup if weren't configured with
one, e.g. if we're upgrading an existing cluster and begin omitting node
ID from the node config.

andrwng committed:
cluster_discovery: add an abort source

andrwng commented on src/v/cluster/controller.cc:
```diff
@@ -469,39 +477,91 @@ ss::future<> controller::stop() {
-        // The controller log has already been written to
+ss::future<>
+controller::cluster_creation_hook(const bool local_node_is_seed_server) {
+    if (config::node().empty_seed_starts_cluster()) {
```
I'm surprised that this has the behavior of not writing a cluster UUID. IMO supporting "legacy" doesn't mean maintaining the exact previous behavior -- we just need to be able to allow for configurations with an empty seed server.

I think ideally, the only thing `empty_seed_starts_cluster` changes is the set of seed servers returned from `initial_seed_brokers()`

dlex replied:
Hmmm right we've been discussing that last week, and this is what is left from my original perspective. I think you're right about it but I just will think twice and will do the command driven bootstrap in both cases. In particular that would mean the original way of creating the initial user will go away, and even with `empty_seed_starts_cluster`, there will be `bootstrap_cluster_cmd` instead of `create_user_cmd`. That should work, though.

dlex committed:
cluster: cluster_bootstrap_info implemented

Client: parallel querying of all seed_servers w/o timeout, until
results are obtained from all peer seed servers.
Verify that both versions and configurations match
Server: supply data
initial_seed_brokers() now returns a future

andrwng commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -79,24 +82,26 @@ ss::future<node_id> cluster_discovery::determine_node_id() {
     if (get_node_index_in_seed_servers()) {
-        // TODO: return the discovered nodes plus this node
-        return {make_self_broker(config::node())};
+        std::vector<model::broker> seed_brokers
+          = co_await request_seed_brokers();
+        seed_brokers.push_back(make_self_broker(config::node()));
+        co_return std::move(seed_brokers);
```
I intended the verification code to be here instead. It seems more fail-safe if we do it here. Doing the validations up front means there's no risk of accidentally exposing a Raft endpoint with a node ID it shouldn't have.

andrwng replied:
Moving the validation here means we also don't need to plug cluster UUID into the controller.

dlex replied:
One validation moved to the `cluster_discovery` ctor because it basically applies to node config, the other one discarded.

andrwng commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -140,6 +145,129 @@ cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+    cluster_bootstrap_info_reply reply;
+    co_await ss::repeat(
+      ss::coroutine::lambda([&reply, addr]() -> ss::future<ss::stop_iteration> {
+          auto reply_result = co_await do_with_client_one_shot<
```
I think we need to try-catch this, in case any servers are down.

andrwng commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -140,6 +145,129 @@ cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+        co_return;
+    });
+
+    bool failed = false;
+    std::vector<model::broker> seed_brokers;
+    seed_brokers.reserve(peers.size());
+    for (auto& peer : peers) {
+        if (
```
nit: possible simplification, maybe we can generate a single reply (for the most part they should all be identical) and check if each node's responses matches what we expected.

dlex replied:
Yes we can, however I just wanted to log precise error diagnostics here in case of any mismatch. It is not possible by just comparing replies.

andrwng commented on src/v/cluster/tests/cluster_test_fixture.h:
```diff
@@ -114,9 +118,12 @@ class cluster_test_fixture {
       int coproc_supervisor_port_base = 43189,
-      configure_node_id use_node_id = configure_node_id::yes) {
+      configure_node_id use_node_id = configure_node_id::yes,
+      empty_seed_starts_cluster empty_seed_starts_cluster_val
```
nit: we should probably distinguish between setting the `empty_seed_starts_cluster` config and configuring with empty seeds.

andrwng reviewed (commented):
Some thoughts so far.

Also sorry but I rebased the node ID assignment branch again -- rebased this branch here https://github.com/andrwng/redpanda/tree/cluster_bootstrap
